### PR TITLE
Use slog in place of waterlog

### PR DESCRIPTION
**Pull request by livingsilver94**
slog is the new structured logging library from the Go project itself. It is available as an experimental package for Go 1.20 and was upstreamed in Go 1.21. This makes it a perfect candidate for our tooling, because:

- it won't bit-rot
- we can use a human-appealing Handler for CLI, and e.g. a JSON Handler on our builders, without changing the code

This PR only replaces waterlog with slog, using its TextHandler Handler. TextHandler is nowhere near the look of waterlog, but a custom Handler mimicking the old look is in the works and will come with a different PR in the near future.

### CODE REVIEW

**Comment by livingsilver94** [comment]
@silkeh would you mind rebasing #13 after this PR is accepted so that we don't keep a miles-long backlog? :) Ideally I would like to see usysconf back amongst the living in two weeks.

**Review by EbonJaeger — DISMISSED** [review]
LGTM, though the changes in `config.load.go` look like a bit more than just changing out the logger. Change scope is something to watch out for in the future, because it makes it harder to review and harder to debug later to figure out what changed caused an issue.

**Comment by livingsilver94** [comment]
I agree with you, but the code was way more duplicated than I could tolerate 😬 I can split the commit anyway.

File: cli/list.go [hunk 1]
```diff
@@ -17,8 +17,8 @@ package cli
 import (
 	"fmt"
 
-	log "github.com/DataDrake/waterlog"
 	"github.com/getsolus/usysconf/config"
+	"golang.org/x/exp/slog"
```
Comment (silkeh) [inline cli/list.go]:
```suggestion
	"log/slog"
```

Idem for all other occurrences.

File: cli/list.go [hunk 2]
```diff
@@ -28,7 +28,7 @@ func (l list) Run(flags GlobalFlags) error {
 	if err != nil {
 		return fmt.Errorf("failed to load triggers: %w", err)
 	}
-	log.Info("Available triggers:\n\n")
+	slog.Info("Available triggers:\n\n")
```
Comment (silkeh) [inline cli/list.go]: Handling of newlines depends on the handler: [tint keeps them in the middle of the line](https://go.dev/play/p/_TQS-9kq5Fi), but the [text handler escapes them](https://go.dev/play/p/WyrTmJMvzKQ). Maybe we should just remove all trailing newlines?

```suggestion
	slog.Info("Available triggers:")
```

File: config/load.go [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Copyright © 2019-2020 Solus Project
+// Copyright © 2019-Present Solus Project
```
Comment (silkeh) [inline config/load.go]: A copyright header indicates the start of copyright, so you know when it enters public domain. `Present` is therefore not a valid value. The rule is that you update the last year every time you edit a file significantly.

File: config/load.go [hunk 4]
```diff
@@ -16,32 +16,29 @@ package config
...
+	"golang.org/x/exp/slog"
 )
 
-// Load reads in all of the trigger files in a directory
-func Load(path string) (tm triggers.Map, err error) {
-	tm = make(triggers.Map)
-	entries, err := ioutil.ReadDir(path)
+// Load reads in all of the trigger files in a directory.
+func Load(path string) (triggers.Map, error) {
+	logger := slog.Default().With("path", path)
```
Comment (silkeh) [inline config/load.go]:
```suggestion
	logger := slog.With("path", path)
```

File: state/map.go [hunk 5]
```diff
@@ -110,7 +110,7 @@ func (m Map) Search(paths []string) Map {
 		search = "^" + strings.ReplaceAll(search, string(filepath.Separator), "\\"+string(filepath.Separator))
 		regex, err := regexp.Compile(search)
 		if err != nil {
-			log.Warnf("Could not convert path to regex: %s\n", path)
+			slog.Error("Could not convert to regex", "path", path)
```
Comment (silkeh) [inline state/map.go]: This can probably stay a warning.

File: state/map.go [hunk 6]
```diff
@@ -133,7 +133,7 @@ func (m Map) Exclude(patterns []string) Map {
 		exclude := strings.ReplaceAll(pattern, "*", ".*")
 		regex, err := regexp.Compile(exclude)
 		if err != nil {
-			log.Warnf("Could not convert pattern to regex: %s\n", pattern)
+			slog.Error("Could not convert to regex", "pattern", pattern)
```
Comment (silkeh) [inline state/map.go]: Idem

**Review by silkeh — CHANGES_REQUESTED** [review]
Great work! I have a few relatively minor comments, but it looks fine otherwise :slightly_smiling_face:

File: config/load.go (hunk 3 above)
Comment (EbonJaeger) [reply config/load.go]: No one can seem to agree on this; other large projects (I think Fedora) says that the dates are completely irrelevant, and they have a lot of lawyers.

File: cli/list.go (hunk 1 above)
Comment (livingsilver94) [reply cli/list.go]: This is pure personal convenience. Fedora so far ships Go 1.20 which hasn't slog in its standard library. I think it's harmless to keep it like that for a while.

File: config/load.go (hunk 3 above)
Comment (livingsilver94) [reply config/load.go]: This was my source: https://hynek.me/til/copyright-years/

Comment (silkeh) [reply config/load.go]: I think the thing no one can seem to agree on is if a statement like this is actually required. As far as I know, it's not.

This means that the copyright statement in it's entirety is informative: the dates in the copyright statement can be used to inform receivers of the source code in the future of the copyright status. This is more relevant when the source is not available in a Git repo with dates and authors.

I'm all for just removing the years entirely (or the last year, at least) by the way, it's just that having *present* there defeats the purpose of the years in the copyright statement.

Comment (livingsilver94) [reply config/load.go]: React used to use "present" before removing the year range altogether. Both are probably sort of valid, but at this point let's embrace your suggestion and the React way.

**Comment by livingsilver94** [comment]
All sorted, I also squashed the commits.

**Review by silkeh — APPROVED** [review]
LGTM, thanks!

**Pull request merged**